src/storage-controller/src/collection_mgmt.rs

```diff
@@ -192,13 +347,476 @@ where
+        &mut self,
+        batch: Vec<(
+            Vec<(Row, i64)>,
+            oneshot::Sender<Result<(), StorageError<T>>>,
+        )>,
+    ) -> Result<(), ()> {
+        // To rate limit appends to persist we add artifical latency, and will
+        // finish no sooner than this instant.
+        let batch_duration_ms = match self.id {
+            GlobalId::User(_) => {
+                Duration::from_millis(self.user_batch_duration_ms.load(Ordering::Relaxed))
+            }
+            // For non-user collections, always just use the default.
+            _ => STORAGE_MANAGED_COLLECTIONS_BATCH_DURATION_DEFAULT,
+        };
+        let use_batch_now = Instant::now();
+        let min_time_to_complete = use_batch_now + batch_duration_ms;
+
+        tracing::debug!(
+            ?use_batch_now,
+            ?batch_duration_ms,
+            ?min_time_to_complete,
+            "batch duration",
+        );
+
+        // Reset the interval which is used to periodically bump the uppers
+        // because the uppers will get bumped with the following update. This
+        // makes it such that we will write at most once every `interval`.
+        //
+        // For example, let's say our `DEFAULT_TICK` interval is 10, so at `t +
+        // 10`, `t + 20`, ... we'll bump the uppers. If we receive an update at
+        // `t + 3` we want to shift this window so we bump the uppers at `t +
+        // 13`, `t + 23`, ... which resetting the interval accomplishes.
+        self.upper_tick_interval.reset();
+
+        let (rows, responders): (Vec<_>, Vec<_>) = batch.into_iter().unzip();
+
+        let rows = rows.into_iter().flatten().collect_vec();
+
+        self.to_write.extend(rows.iter().cloned());
+        self.desired.extend(rows.iter().cloned());
+
+        // TODO: Maybe don't do it every time?
+        consolidation::consolidate(&mut self.desired);
+        consolidation::consolidate(&mut self.to_write);
+
+        self.write_to_persist(responders).await?;
+
+        // Wait until our artificial latency has completed.
+        //
+        // Note: if writing to persist took longer than `DEFAULT_TICK` this
+        // await will resolve immediately.
+        tokio::time::sleep_until(min_time_to_complete).await;
+
+        Ok(())
+    }
+
+    /// Attempt to write what is currently in [Self::to_write] to persist,
+    /// retrying and re-syncing to persist when necessary, that is when the
+    /// upper was not what we expected.
+    async fn write_to_persist(
+        &mut self,
+        responders: Vec<oneshot::Sender<Result<(), StorageError<T>>>>,
+    ) -> Result<(), ()> {
+        // We'll try really hard to succeed, but eventually stop.
+        //
+        // Note: it's very rare we should ever need to retry, and if we need to
+        // retry it should only take 1 or 2 attempts. We set `max_tries` to be
+        // high though because if we hit some edge case we want to try hard to
+        // commit the data.
```

Comment:
Is that reasoning still the same now that we explicitly account for concurrent writers? I assume it is as long as we don't actually have them, but we might run into surprises when we start having them and forget about this code that limits the number of retries we are attempting.

Reply:
I think even with that it's good to limit the retries: we still don't shut down the write task, so the next time something in `desired` changes we would get a full set of retries again. That seems good, what do you think?

Reply:
Fine with me!